src/datahike/db.cljc

```diff
@@ -294,25 +312,305 @@
+        from (components->pattern db index-type cs e0 tx0)
+        to (datom emax nil nil txmax)]
+    (concat
+     (-> (concat (-slice  index from to index-type)
+                 (-slice temporal-index from to index-type))
+         vec
+         rseq))))
+
+(defn temporal-index-range [^DB db current-db attr start end]
+  (when-not (indexing? db attr)
+                  (raise "Attribute" attr "should be marked as :db/index true" {}))
+                (validate-attr attr (list '-index-range 'db attr start end) db)
+                (let [from (resolve-datom current-db nil attr start nil e0 tx0)
+                      to (resolve-datom current-db nil attr end nil emax txmax)]
+                  (concat
+                   (-slice (get db :avet) from to :avet)
+                   (-slice (get db :temporal-avet) from to :avet))))
+
+(defrecord-updatable HistoricalDB [origin-db]
+  #?@(:cljs
+      [IEquiv (-equiv [db other] (equiv-db db other))
+       ISeqable (-seq [db] (-datoms db :eavt []))
+       ICounted (-count [db] (count (-datoms db :eavt [])))
+       IPrintWithWriter (-pr-writer [db w opts] (pr-db db w opts))
+
+       IEmptyableCollection (-empty [_] (throw (js/Error. "-empty is not supported on FilteredDB")))
+       IEmptyableCollection (-empty [_] (throw (js/Error. "-empty is not supported on FilteredDB")))
+
+       ILookup (-lookup ([_ _] (throw (js/Error. "-lookup is not supported on FilteredDB")))
+                        ([_ _ _] (throw (js/Error. "-lookup is not supported on FilteredDB"))))
+
+       IAssociative (-contains-key? [_ _] (throw (js/Error. "-contains-key? is not supported on FilteredDB")))
+       (-assoc [_ _ _] (throw (js/Error. "-assoc is not supported on FilteredDB")))]
+      :clj
+      [clojure.lang.IPersistentCollection
+       (count [db] (count (-datoms db :eavt [])))
+       (equiv [db o] (equiv-db db o))
+       (cons [db [k v]] (throw (UnsupportedOperationException. "cons is not supported on CurrentDB")))
+       (empty [db] (throw (UnsupportedOperationException. "empty is not supported on CurrentDB")))
+
+       clojure.lang.Seqable (seq [db] (-datoms db :eavt []))
+
+       clojure.lang.ILookup (valAt [db k] (throw (UnsupportedOperationException. "valAt/2 is not supported on CurrentDB")))
+       (valAt [db k nf] (throw (UnsupportedOperationException. "valAt/3 is not supported on CurrentDB")))
+       clojure.lang.IKeywordLookup (getLookupThunk [db k]
+                                     (throw (UnsupportedOperationException. "getLookupThunk is not supported on CurrentDB")))
+
+       clojure.lang.Associative
+       (containsKey [e k] (throw (UnsupportedOperationException. "containsKey is not supported on CurrentDB")))
+       (entryAt [db k] (throw (UnsupportedOperationException. "entryAt is not supported on CurrentDB")))
+       (assoc [db k v] (throw (UnsupportedOperationException. "assoc is not supported on CurrentDB")))])
+
```

Comment:
In the above, "FilteredDB" should be replaced by "HisotricalDB" in the cljs part and the same for 'CurrentDB' in the clj part.